### ISSUE TITLE
Add LDAP/AD integration instruction

### ISSUE THREAD

**Issue by jakalkhalili** [jenkinsci/kubernetes-operator#133]
As written here - https://github.com/jenkinsci/kubernetes-operator/issues/120, there is need to add LDAP/AD integration instruction in customization docs:

https://github.com/jenkinsci/kubernetes-operator/blob/master/website/content/en/docs/Getting%20Started/v0.2.0/customization.md 

and here 

https://github.com/jenkinsci/kubernetes-operator/blob/master/website/content/en/docs/Getting%20Started/v0.1.1/customization.md

Issues that may have relevant info: #208 #191 #192

**Comment by yoshi314** [comment]
What i do is alter the created user password for the jenkins-operator to an account that exists in LDAP and apply a following yaml file via JCaSC : 

```
  1-configure-ldap.yaml: |
    # ldap - ldap configuration must be provided
    jenkins:
      systemMessage: "LDAP configured"
      securityRealm:
        ldap:
          configurations:
            - server: "ldap://ldap.server"
              groupMembershipStrategy:
                fromUserRecord:
                  attributeName: "memberOf"
              rootDN: "dc=some,dc=company,dc=com" # mandatory
              managerDN: "CN=ldapmanager,OU=sysusers,OU=Users,OU=sysaccounts,DC=some,DC=company,DC=com"
              managerPasswordSecret: "sekrit"
              userSearchBase: 
              userSearch: "sAMAccountName={0}"
              groupSearchBase: 
              groupSearchFilter: "sAMAccountName={0}"
```

hope this helps

**Comment by pawelprazak** [comment]
might want to also take look at #192 for things to clarify in the docs

**Comment by bechampion** [comment]
The documentation links above return 404 ,  I've been looking at the code today , i can see where the random credentials are generated etc.  Still quite unclear what happens after install AD or LDAP plugins on jenkins.
Once you install LDAP or AD plugins , are you supposed to change the kube-secret that contain the credentials?  that will trigger an userpasswd hash mismatch and the jenkins pod will be recreated.
But then i seem to end up in a reconcile loop ,  is like the jenkins controller doesn't really pick up the credentials from the secret anymore?

Has anybody had some success with this?

**Comment by tomaszsek** [comment]
Instructions:
- get the access to the LDAP
- create user for the operator in LDAP
- save the username and password to the credentials secret
- configure LDAP in Jenkins through Groovy scripts or JasC
- give admin access to the operator user through Groovy scripts or JasC

**Comment by yoshi314** [comment]
> The documentation links above return 404 , I've been looking at the code today , i can see where the random credentials are generated etc. Still quite unclear what happens after install AD or LDAP plugins on jenkins.
> Once you install LDAP or AD plugins , are you supposed to change the kube-secret that contain the credentials? that will trigger an userpasswd hash mismatch and the jenkins pod will be recreated.
> But then i seem to end up in a reconcile loop , is like the jenkins controller doesn't really pick up the credentials from the secret anymore?
> 
> Has anybody had some success with this?

i replaced the jenkins-operator credential with one that has correct login and password for ldap. The token has to be left as-is, so you'll have to dump the secret first , edit it and reapply.

**Comment by bechampion** [comment]
Hey all ,  i have it working , thanks so much for the help ,  two thigns to notice , it worked for me with these plugins version:
```
  - name: configuration-as-code-support    
    version: "1.18"
    version: "1.21.0"    
  - name: ldap    
    version: "1.21"```
The other thing to say is once you change the password on the secret there's a reconciliation that happens and the pods get recreated.

**Comment by BogdanSorlea** [comment]
> What i do is alter the created user login+password for the jenkins-operator to an account that exists in LDAP and apply a following yaml file via JCaSC :

How exactly do you alter the jenkins-operator password to the one you set-up in LDAP (if I understand this correctly)? @yoshi314 

> save the username and password to the credentials secret

It is unclear what credentials secret and how this gets plugged in for jenkins-operator to use @tomaszsek 

if you guys could clarify, including with examples, it would help a lot. what I understand is that I create in our LDAP a `jenkins-operator` user with a "fixed" password, password which gets written to some Secret k8s resource that somehow gets plugged in such that the operator pod (and I guess the jenkins pod as well, until LDAP connection is set-up) knows about it and uses it. but it's very unclear how this is achieved. examples welcomed.

**Comment by tomaszsek** [comment]
@BogdanSorlea I've updated the comment with the secret example https://github.com/jenkinsci/kubernetes-operator/issues/133#issuecomment-555976832. We will update the documentation with the step by step instruction.

**Comment by BogdanSorlea** [comment]
cool, thanks. I also found in the meantime the secret via `kubectl get secrets`, but this gives better clarity.

**Comment by BogdanSorlea** [comment]
the last step does not seem to work,
```
WARNING: Configuration-as-Code can't handle type class hudson.security.GlobalMatrixAuthorizationStrategy
```
code being,
```
jenkins:
  authorizationStrategy:
    globalMatrix:
      permissions:
        - "Overall/Administer:authenticated,jenkins-operator,SYSTEM"
```

could you point me to the updated documentation, when it will be available?

**Comment by BogdanSorlea** [comment]
I hereby confirm that this worked - https://github.com/jenkinsci/kubernetes-operator/issues/133#issuecomment-555976832

**Comment by yoshi314** [comment]
@BogdanSorlea  i've setup the permissions in web ui, and then exported them via configuration as code plugin. I used per-project permissions.

Here is a relevant snippet 

```
 3-permission-matrix.yaml: |
    jenkins:
      authorizationStrategy:
          projectMatrix:
            permissions:
            - "Agent/Build:authenticated"
            - "Agent/Build:jenkins-operator-dev"
            - "Agent/Build:m15kowal"
            - "Agent/Configure:authenticated"
            - "Agent/Configure:jenkins-operator-dev"
            - "Agent/Configure:m15kowal"
            - "Agent/Connect:authenticated"
            - "Agent/Connect:jenkins-operator-dev"
            - "Agent/Connect:m15kowal"
            - "Agent/Create:authenticated"
            - "Agent/Create:jenkins-operator-dev"
            - "Agent/Create:m15kowal"
            - "Agent/Delete:authenticated"
            - "Agent/Delete:jenkins-operator-dev"
            - "Agent/Delete:m15kowal"
            - "Agent/Disconnect:authenticated"
            - "Agent/Disconnect:jenkins-operator-dev"
            - "Agent/Disconnect:m15kowal"
            - "Credentials/Create:authenticated"
            - "Credentials/Create:jenkins-operator-dev"
            - "Credentials/Create:m15kowal"
            - "Credentials/Delete:authenticated"
            - "Credentials/Delete:jenkins-operator-dev"
(....)
```

**Comment by tomaszsek** [comment]
Please remember that user for the jenkins operator have to have admin rights.

**Comment by yoshi314** [comment]
Yes, the exported yaml was a bit long, so i cut it short here.

**Comment by apogrebnyak** [comment]
There is a problem with using `jenkins-operator-credentials-<cr-name>` secret.

And the problem is that it stores 2 generated secret values for `token` and `tokenCreationTime`.

I am using terraform for kubernetes secret management.  On initial TF deploy everything is fine, however on consequent deploys, when TF tries to reconcile `jenkins-operator-credentials-<cr-name>`, both `token` and `tokenCreationTime` get deleted, with disastrous consequences (basically breaking operator->cr communication channel).

I propose moving token and tokenCreationTime to its own secret, so we can specify custom credential secrets in case of LDAP configuration.

**Comment by VGerris** [comment]
since it seems when LDAP is down, login is impossible, a way to deal with that would be appreciated too. And more preferable of course, just internal admin users working

**Comment by Bakies** [comment]
FWIW Since I was having trouble following the instructions in https://github.com/jenkinsci/kubernetes-operator/issues/133#issuecomment-562056816 I'll detail what I did. I was having trouble with the operator overriding the LDAP settings I had in my CasC yaml when it ran the script in the configmap jenkins-operator-init-configuration-crname. I ended up just mounting an empty dir in the jenkins cr with the path `/var/lib/jenkins/operatorUserCreated` so the script doesn't change the security realm settings.
